File: src/types/widget-data.ts
/**
 * ホームウィジェット（Widget Extension）に表示するデータの型定義。
 *
 * 背景: メインアプリと Widget Extension は App Groups UserDefaults 経由で
 * JSON 文字列としてデータを共有する。この型は JS 側・Swift 側（SharedTypes.swift）
 * の両方で同じ構造を維持する必要がある。
 *
 * ライフサイクル: buildWidgetData() で組み立て → syncWidgetData() で UserDefaults に書き出し
 * → Widget Extension の TimelineProvider が読み取り → WidgetKit で表示
 */

export interface WidgetTodo {
  readonly id: string;
  readonly title: string;
  readonly completed: boolean;
}

export interface WidgetData {
  readonly nextAlarm: {
    readonly time: string; // "HH:mm"
    readonly enabled: boolean;
    readonly label: string; // 曜日ラベル（例: "月"）
  } | null;
  readonly session: {
    readonly todos: readonly WidgetTodo[];
    readonly snoozeFiresAt: string | null;
    readonly progress: { readonly completed: number; readonly total: number };
  } | null;
  readonly streak: {
    readonly currentStreak: number;
    readonly lastGrade: string; // DailyGrade
  };
  readonly updatedAt: string;
}

File: targets/widget/WidgetData.swift
import Foundation

// MARK: - Widget Data
//
// ホームウィジェットに表示するデータ。メインアプリの buildWidgetData()
// (src/domain/widget-data.ts) が組み立てた JSON を App Groups UserDefaults の
// "widget-data" キーから読み取る。
//
// 重要: プロパティ構成は src/types/widget-data.ts の WidgetData と完全に一致させること。
// 片方だけ変更すると JSON デコードに失敗してウィジェットが空表示になる。

struct WidgetData: Codable {
  let nextAlarm: NextAlarm?
  let session: SessionData?
  let streak: Streak
  let updatedAt: String

  struct NextAlarm: Codable {
    let time: String  // "HH:mm"
    let enabled: Bool
    let label: String  // 曜日ラベル（例: "月"）
  }

  struct WidgetTodo: Codable, Identifiable {
    let id: String
    let title: String
    let completed: Bool
  }

  struct Progress: Codable {
    let completed: Int
    let total: Int
  }

  struct SessionData: Codable {
    let todos: [WidgetTodo]
    let snoozeFiresAt: String?
    let progress: Progress
  }

  struct Streak: Codable {
    let currentStreak: Int
    let lastGrade: String
  }
}

/// App Groups 経由で WidgetData を読み取るストア。
///
/// App Group ID は AlarmKitService.ts / expo-target.config.js と同じ値を使用する。
/// 変更時は 3 箇所すべてを揃えること。
enum WidgetDataStore {
  static let appGroupId = "group.com.tktcorporation.goodmorning"
  static let storageKey = "widget-data"

  static func load() -> WidgetData? {
    guard let defaults = UserDefaults(suiteName: appGroupId),
      let json = defaults.string(forKey: storageKey),
      let data = json.data(using: .utf8)
    else {
      return nil
    }
    return try? JSONDecoder().decode(WidgetData.self, from: data)
  }
}
